fableous-fe/src/components/Canvas.tsx

```diff
@@ -18,417 +18,609 @@ import { ControllerRole, ToolMode, WSMessage, WSMessageType } from "../Data";
-      ];
-      console.log(startX, startY);
-      const setPixel = (pixel: number, rgb: number[]) => {
-        [
-          image.data[pixel],
-          image.data[pixel + 1],
-          image.data[pixel + 2],
-          image.data[pixel + 3],
-        ] = rgb;
-      };
-      const checkPixel = (pixel: number, rgb: number[]) => {
-        return (
-          image.data[pixel] === rgb[0] &&
-          image.data[pixel + 1] === rgb[1] &&
-          image.data[pixel + 2] === rgb[2] &&
-          image.data[pixel + 3] === rgb[3]
-        );
-      };
-      const colorRGB = convHEXtoRGBA(targetColor);
-      const stack = [[startX, startY]];
-      if (checkPixel(startPixel, colorRGB)) return;
-      while (stack.length) {
-        // eslint-disable-next-line prefer-const
-        let [x, y] = stack.pop() || [startX, startY];
-        let pixel = (y * width + x) * 4;
-        while (y-- > 0 && checkPixel(pixel, startColor)) {
-          pixel -= width * 4;
+    const convHEXtoRGBA = (hex: string) => {
+      const result =
+        /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
+      return result
+        ? [
+            parseInt(result[1], 16),
+            parseInt(result[2], 16),
+            parseInt(result[3], 16),
+            parseInt(result[4], 16),
+          ]
+        : [0, 0, 0, 0];
+    };
+
+    const getTextBounds = useCallback(
+      (x: number, y: number, text: string, fontSize: number) => {
+        const bounds = canvasRef.current?.getContext("2d")?.measureText(text);
+        if (bounds) {
+          return [
+            x - bounds.width / 2,
+            y - (fontSize * SCALE) / 2,
+            x + bounds.width / 2,
+            y + (fontSize * SCALE) / 2,
+          ];
         }
-        pixel += width * 4;
-        y++;
-        let expandLeft = false;
-        let expandRight = false;
-        while (y++ < height - 1 && checkPixel(pixel, startColor)) {
-          setPixel(pixel, colorRGB);
-          if (x > 0) {
-            if (checkPixel(pixel - 4, startColor) && !expandLeft) {
-              stack.push([x - 1, y]);
-              expandLeft = true;
-            } else {
-              expandLeft = false;
-            }
+        return [0, 0, 0, 0];
+      },
+      [canvasRef]
+    );
+
+    const placePaint = useCallback(
+      (
+        x1: number,
+        y1: number,
+        x2: number,
+        y2: number,
+        targetColor: string,
+        targetWidth: number
+      ) => {
+        const ctx = canvasRef.current.getContext(
+          "2d"
+        ) as CanvasRenderingContext2D;
+        ctx.beginPath();
+        ctx.lineCap = "round";
+        ctx.lineJoin = "round";
+        ctx.strokeStyle = targetColor;
+        ctx.lineWidth = targetWidth;
+        ctx.moveTo(x1, y1);
+        ctx.lineTo(x2, y2);
+        ctx.closePath();
+        ctx.stroke();
+        if (role !== ControllerRole.Hub) {
+          const [normX1, normY1] = scaleDownXY(x1, y1);
```

Comment:
perhaps scaling down and up logic before send and after recv web socket events respectively can be put into a wrapper of websocket reference(?), super low priority, could just be captured to a task if u want

for that, another file for Canvas will be needed, perhaps a folder can be used for this? as this is a compex component.

Helper funcs such as translateXY and scaleDownXY could then be put into separate file, making canvasRef as argument so its stateless (or whatever the term is), again super low priority, could just be captured to a task if u want

Reply:
got it